### PR TITLE
[DF] Avoid jitting a very large function body, which is slow

### PR DESCRIPTION
**Pull request by eguiraud**
Instead, split the code RDataFrame jits every 1k newlines and jit
in chunks. This avoids an explosion in the time required by
optimizer passes, which seem to not deal well with very large
function bodies.

See https://github.com/root-project/root/issues/9312 and
https://github.com/root-project/root/issues/7604 for more discussion.

@hahnjo can you confirm this helps? It brings jitting times down to reasonable (<1m times) for the cases I tested.

### CODE REVIEW

**Review by hahnjo — APPROVED** [review]
Well, I only have the one test case you provided in #9312 and I can confirm this now finishes in 30s instead of 13m on my machine, but I guess you also tested this. Other than that, we only need to make sure that we never have statements spanning more than one line, or the splitting will go horribly wrong...

**Comment by eguiraud** [comment]
> we only need to make sure that we never have statements spanning more than one line

we control the generation of that code, so that should not happen/tests will break noisily and we can fix them if it happens.

**Comment by hahnjo** [comment]
> > we only need to make sure that we never have statements spanning more than one line
> 
> we control the generation of that code, so that should not happen/tests will break noisily and we can fix them if it happens.

Yes, they will fail *if* they exceed the threshold of 1000 lines :smiley:

File: tree/dataframe/src/RDFUtils.cxx [hunk 1]
```diff
@@ -333,17 +333,31 @@ Long64_t InterpreterCalc(const std::string &code, const std::string &context)
...
+      gInterpreter->Calc(subs.c_str(), &errorCode);
+      if (errorCode != TInterpreter::EErrorCode::kNoError) {
+         std::string msg = "\nAn error occurred during just-in-time compilation";
+         if (!context.empty())
+            msg += " in " + context;
+         msg +=
+            ". The lines above might indicate the cause of the crash\nAll RDF objects that have not run their event "
+            "loop yet should be considered in an invalid state.\n";
+         throw std::runtime_error(msg);
+      }
```
Comment (Axel-Naumann) [inline tree/dataframe/src/RDFUtils.cxx:357]: This looks like a great candidate for a dedicated function. Maybe even splitting the lines into a `vector<string>` first, then then calling that function on each element.

**Review by Axel-Naumann — APPROVED** [review]
LGTM either way!

File: tree/dataframe/src/RDFUtils.cxx (hunk 1 above)
Comment (eguiraud) [reply tree/dataframe/src/RDFUtils.cxx:357]: funny you'd say that, as that was precisely the body of the function before these changes! :smile: 

I made it a lambda. Everything still fits in half a screen.

**Comment by eguiraud** [comment]
@hahnjo i just double-checked. in RDF we generate one long line per function call, no newlines in between. a sample:

```
ROOT::Internal::RDF::JitDefineHelper<ROOT::Internal::RDF::DefineTypes::RDefineTag>(R_rdf::lambda0, new const char*[0]{}, 0, "x0", reinterpret_cast<ROOT::Detail::RDF::RLoopManager*>(0x559209672780), reinterpret_cast<std::weak_ptr<ROOT::Detail::RDF::RJittedDefine>*>(0x559209e0d6c0), reinterpret_cast<ROOT::Internal::RDF::RBookedDefines*>(0x55920ad807e0), reinterpret_cast<std::shared_ptr<ROOT::Detail::RDF::RNodeBase>*>(0x559209673860));
ROOT::Internal::RDF::JitFilterHelper(R_rdf::lambda1, new const char*[1]{"x0"}, 1, "", reinterpret_cast<std::weak_ptr<ROOT::Detail::RDF::RJittedFilter>*>(0x55920ac84440), reinterpret_cast<std::shared_ptr<ROOT::Detail::RDF::RNodeBase>*>(0x55920ac88e00),reinterpret_cast<ROOT::Internal::RDF::RBookedDefines*>(0x55920ad84cf0));
ROOT::Internal::RDF::CallBuildAction<ROOT::Internal::RDF::ActionTags::Sum, float>(reinterpret_cast<std::shared_ptr<ROOT::Detail::RDF::RNodeBase>*>(0x55920afc2c30), new const char*[1]{"x0"}, 1, 1, reinterpret_cast<shared_ptr<double>*>(0x55920afc2c10), reinterpret_cast<std::weak_ptr<ROOT::Internal::RDF::RJittedAction>*>(0x55920afc2c50), reinterpret_cast<ROOT::Internal::RDF::RBookedDefines*>(0x55920ad84e30));ROOT::Internal::RDF::JitDefineHelper<ROOT::Internal::RDF::DefineTypes::RDefineTag>(R_rdf::lambda0, new const char*[0]{}, 0, "x1", reinterpret_cast<ROOT::Detail::RDF::RLoopManager*>(0x559209672780), reinterpret_cast<std::weak_ptr<ROOT::Detail::RDF::RJittedDefine>*>(0x55920accbb40), reinterpret_cast<ROOT::Internal::RDF::RBookedDefines*>(0x55920acd4020), reinterpret_cast<std::shared_ptr<ROOT::Detail::RDF::RNodeBase>*>(0x55920accb7e0));
ROOT::Internal::RDF::JitFilterHelper(R_rdf::lambda1, new const char*[1]{"x1"}, 1, "", reinterpret_cast<std::weak_ptr<ROOT::Detail::RDF::RJittedFilter>*>(0x55920accfb50), reinterpret_cast<std::shared_ptr<ROOT::Detail::RDF::RNodeBase>*>(0x55920accfaa0),reinterpret_cast<ROOT::Internal::RDF::RBookedDefines*>(0x55920accfd60));
```

**Pull request merged**